### ISSUE TITLE
Read mzML file - bad lexical cast: source type value could not be interpreted as target

### ISSUE THREAD

**Issue by lgatto** [lgatto/MSnbase#383]
I have this [mzML file](https://transvol.sgsi.ucl.ac.be/download.php?id=8eb7f8a884816181) that was written using `writeMSData` after processing a raw data, itself read in using `readMSData`. The processing included spectra merging and coercing a `Spectra` object to the `MSnExp` that produced the test file. 

When I not try to read it back in:
```r
> readMSData("~/tmp/test.mzML", mode = "onDisk")
Error in object@backend$getAllScanHeaderInfo() : 
  bad lexical cast: source type value could not be interpreted as target
> readMSData("~/tmp/test.mzML", mode = "inMemory")
Error in object@backend$getAllScanHeaderInfo() : 
  bad lexical cast: source type value could not be interpreted as target
```
@jotsetung - any idea?

This is with
```r
> sessionInfo()
R version 3.5.1 Patched (2018-11-02 r75540)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Manjaro Linux

Matrix products: default
BLAS: /usr/lib/libblas.so.3.8.0
LAPACK: /usr/lib/liblapack.so.3.8.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] plotly_4.8.0        ggplot2_3.1.0       DT_0.5             
 [4] xcms_3.4.1          MSnbase_2.9.2       ProtGenerics_1.14.0
 [7] S4Vectors_0.20.1    mzR_2.16.0          Rcpp_1.0.0         
[10] BiocParallel_1.16.0 Biobase_2.42.0      BiocGenerics_0.28.0

loaded via a namespace (and not attached):
 [1] lattice_0.20-35        tidyr_0.8.2            assertthat_0.2.0      
 [4] digest_0.6.18          foreach_1.4.4          R6_2.3.0              
 [7] plyr_1.8.4             mzID_1.20.0            httr_1.3.1            
[10] pillar_1.3.0           zlibbioc_1.28.0        rlang_0.3.0.1         
[13] lazyeval_0.2.1         data.table_1.11.8      Matrix_1.2-14         
[16] preprocessCore_1.44.0  splines_3.5.1          htmlwidgets_1.3       
[19] munsell_0.5.0          compiler_3.5.1         pkgconfig_2.0.2       
[22] multtest_2.38.0        pcaMethods_1.74.0      htmltools_0.3.6       
[25] tidyselect_0.2.5       tibble_1.4.2           RANN_2.6              
[28] IRanges_2.16.0         codetools_0.2-15       XML_3.98-1.16         
[31] viridisLite_0.3.0      crayon_1.3.4           dplyr_0.7.8           
[34] withr_2.1.2            MASS_7.3-51.1          grid_3.5.1            
[37] MassSpecWavelet_1.48.0 jsonlite_1.5           gtable_0.2.0          
[40] affy_1.60.0            magrittr_1.5           scales_1.0.0          
[43] ncdf4_1.16             impute_1.56.0          affyio_1.52.0         
[46] doParallel_1.0.14      bindrcpp_0.2.2         limma_3.38.2          
[49] robustbase_0.93-3      RColorBrewer_1.1-2     iterators_1.0.10      
[52] tools_3.5.1            glue_1.3.0             DEoptimR_1.0-8        
[55] purrr_0.2.5            survival_2.43-1        colorspace_1.3-2      
[58] BiocManager_1.30.4     vsn_3.50.0             MALDIquant_1.18       
[61] bindr_0.1.1           
```

**Comment by jotsetung** [comment]
Sounds like some information got lost during the processing. I have a look at it.

**Comment by lgatto** [comment]
Thanks - [here](https://transvol.sgsi.ucl.ac.be/download.php?id=d33f745722f28335)'s also the serialised `MSnExp`.

**Comment by jotsetung** [comment]
My first guess is that `<spectrum index="1" id="scan=NA" defaultArrayLength="4859"> ` is the problem, specifically the `"scan=NA"`.

**Comment by lgatto** [comment]
Indeed:
```r
> scanIndex(res)
$F1.S00002
integer(0)

$F1.S00006
integer(0)

$F1.S00010
integer(0)

$F1.S00014
integer(0)

$F1.S00018
integer(0)

$F1.S00022
integer(0)

$F1.S00026
integer(0)

$F1.S00030
integer(0)

$F1.S00034
integer(0)

$F1.S00038
integer(0)

> test[[1]]@scanIndex
integer(0)
```

Would this be something that isn't properly set in `combineSpectra'?

**Comment by jotsetung** [comment]
I think that is up to the function that is called by `combineSpectra`. That one should ensure that valid spectra are returned. Still, I got the same problem after setting `scanIndex`

**Comment by jotsetung** [comment]
Spotted the problem. The export of `MSnExp` objects is buggy and not all header columns get passed along correctly. I'm working on a fix.

**Comment by jotsetung** [comment]
I made a PR for the fix. Now it is using the `spectrumId` which is provided in `fData`, or it is generating an ID based on the `acquisitionNum`.

**Comment by lgatto** [comment]
Thank you @jotsetung - I guess this should be backported to release, right?

**Comment by lgatto** [comment]
I confirm that #384 fixed my issue, hence closing.

**Comment by jotsetung** [comment]
Ah, yes, please also add that to release. Thanks!

**Comment by lgatto** [comment]
Done. Actually, I was a bit lazy and, as there was another fix (#381) to incorporate, I merged all of master. We now have 2.8.2 and 2.9.2 that are fully sync'ed.
